### DISCUSSION TITLE
MvvmCross.Plugin.Sidebar

### DISCUSSION THREAD

**Discussion by Larhei** [MvvmCross/MvvmCross#4569] [Q&A] [answered]
During migrating to .Net6 and MvvmCross 9 we are facing the issue that the Sidebar Plugin was kicked out of the project.
We think it is because of the Dependency on [Xamarin-Sidebar](https://github.com/Baseflow/Xamarin-Sidebar) that is not ready for .Net6. In case @martijn00 would update to a compatible version... Could the Plugin be brought back?

Or does someone have an alternative approach?

**Comment by Cheesebaron** [comment]
You'd have to ask someone from Baseflow if they'd want to keep their package up to date. But I've chosen to remove it since there is no net6.0 support for in that library. Also it is not a UIKit native navigation paradigm, so I don't really want to support this.

You can always just pour in the code into your own project and have a custom presenter that can support this navigation pattern.

**Comment by martijn00** [comment]
I've released https://www.nuget.org/packages/SidebarNavigation/3.0.0 now

**Reply by Larhei** [reply]
@martijn00 Thank you!

**Comment by Larhei** [comment]
Ok @Cheesebaron... back to the initial question.
Can the Plugin be brought back?
You said: "I don't really want to support this".
Last word on this?

**Comment by martijn00** [comment, marked as answer]
I don't think this should be in MvvmCross. It's easy to pull it out to a new repo and release it from there.

**Reply by Larhei** [reply]
Fair

**Comment by cheasley-hub** [comment]
@Larhei, I'm currently facing a similar challenge while upgrading an old Native Xamarin iOS app to .NET 7. This involves years of missed updates. Have you had any success in creating a custom presenter for this purpose in your app?

**Reply by Larhei** [reply]
@cheasley-hub, sorry for the late reply.

I added the Source of the Plugin to my repository, upgraded the ref to SidebarNavigation V3 and fixed some bugs on my repository.
Not the preferred way to go, but for now it is working. Because of other breaking changes in MVVM I need to revisit it in the near future. But it is not at the top of my technical debt list.

**Reply by qmark** [reply]
@Larhei Do you have any progress on the matter? Or is it possible to see the source code for your solution? I cannot see it in your public repos.

**Comment by Larhei** [comment]
@qmark Sorry to say: Still did not have time to revisit this.
What i did:
1. Checkout Version 8.0.2.
2. Copied SidebarPlugin to my project.
3. Modified Csproj to Ref [nuget](https://www.nuget.org/packages/SidebarNavigation/3.0.0).
4. Upgraded to .net8
5. Changed ref to MVVMCross to nuget Version of MVVMCross.
6. Created CustomMvxSidebarPresenter in my iOS Project
```
namespace MvvmCross.Plugin.Sidebar
{
    using System;
    using MvvmCross.Platforms.Ios.Views;
    using MvvmCross.Plugin.Sidebar.Extensions;
    using MvvmCross.Plugin.Sidebar.Views;
    using UIKit;

    public class CustomMvxSidebarPresenter : MvxSidebarPresenter
    {
        public CustomMvxSidebarPresenter(IUIApplicationDelegate applicationDelegate, UIWindow window)
                    : base(applicationDelegate, window)
        {
            this.SideBarViewController = new CustomMvxSidebarViewController();
        }

        public void DisablePanGesture()
        {
            if (this.SideBarViewController?.LeftSidebarController != null)
            {
                this.SideBarViewController.LeftSidebarController.DisablePanGesture = true;
            }

            if (this.SideBarViewController?.RightSidebarController != null)
            {
                this.SideBarViewController.RightSidebarController.DisablePanGesture = true;
            }
        }

        public void EnablePanGesture()
        {
            if (this.SideBarViewController?.LeftSidebarController != null)
            {
                this.SideBarViewController.LeftSidebarController.DisablePanGesture = false;
            }

            if (this.SideBarViewController?.RightSidebarController != null)
            {
                this.SideBarViewController.RightSidebarController.DisablePanGesture = false;
            }
        }

        protected override MvxNavigationController CreateNavigationController(UIViewController viewController)
        {
            if (viewController == null)
            {
                throw new ArgumentNullException(nameof(viewController));
            }

            return new CustomMvxNavigationController(viewController);
        }

        protected override bool ShowPanelAndPopToRoot(MvxSidebarPresentationAttribute attribute, UIViewController viewController)
        {
            var navigationController = (this.SideBarViewController as MvxSidebarViewController).NavigationController;

            if (navigationController == null)
            {
                return false;
            }

            navigationController.PopToRootViewController(false);
            navigationController.PushViewController(viewController, false);

            return true;
        }

        protected override bool ShowPanelAndResetToRoot(MvxSidebarPresentationAttribute attribute, UIViewController viewController)
        {
            var navigationController = (this.SideBarViewController as MvxSidebarViewController).NavigationController;

            if (navigationController == null)
            {
                return false;
            }

            navigationController.ViewControllers = new[] { viewController };

            switch (attribute.Panel)
            {
                case MvxPanelEnum.Left:
                case MvxPanelEnum.Right:
                    break;
                case MvxPanelEnum.Center:
                default:

                    viewController.ShowMenuButton(this.SideBarViewController, showLeft: false, showRight: false);
                    break;
                case MvxPanelEnum.CenterWithLeft:

                    viewController.ShowMenuButton(this.SideBarViewController, showLeft: false, showRight: false);
                    break;
                case MvxPanelEnum.CenterWithRight:

                    viewController.ShowMenuButton(this.SideBarViewController, showLeft: false, showRight: false);
                    break;
            }

            return true;
        }

        protected override bool ShowPanel(MvxSidebarPresentationAttribute attribute, UIViewController viewController)
        {
            var navigationController = this.SideBarViewController.NavigationController;

            if (navigationController == null)
            {
                return false;
            }

            switch (attribute.Panel)
            {
                case MvxPanelEnum.Left:
                case MvxPanelEnum.Right:
                    break;
                case MvxPanelEnum.Center:
                default:
                    navigationController.PushViewController(viewController, attribute.Animated);
                    break;
                case MvxPanelEnum.CenterWithLeft:
                    navigationController.PushViewController(viewController, attribute.Animated);
                    viewController.ShowMenuButton(this.SideBarViewController, showLeft: false, showRight: false);
                    break;
                case MvxPanelEnum.CenterWithRight:
                    navigationController.PushViewController(viewController, attribute.Animated);
                    viewController.ShowMenuButton(this.SideBarViewController, showLeft: false, showRight: false);
                    break;
            }

            return true;
        }
    }
}

public class CustomMvxSidebarViewController : MvxSidebarViewController
{
    public CustomMvxSidebarViewController()
    {
        this.StatusBarHidden = true;
    }
}
```

7. Added this to my iOS Setup.cs
`
        protected override IMvxIosViewPresenter CreateViewPresenter()
        {
            return new CustomMvxSidebarPresenter((MvxApplicationDelegate)this.ApplicationDelegate, this.Window) { };
        }
`

**Reply by qmark** [reply]
thank you very much! In previous post you also mentioned fixing some bugs, are there a lot of those after .net8 migration?

**Reply by Larhei** [reply]
As of my knowledge it works for me. But we only use it in a Debug Build to have additional menu options.. So you would need to test for yourself. Fixing some bugs where all related to our coding guideline So no technical bugs.

**Reply by cheasley-hub** [reply]
@qmark Did you happen to implement this in your project?

**Reply by qmark** [reply]
Yes, it seems to have no issues on .net8, I migrated pretty quickly

**Reply by cheasley-hub** [reply]
Thank you. I think I finally figured out my own issue with it after circling back to it today. I'm having to update from MvvmCross version 5 to 9 though, so there are more than a few changes that are causing me pain.
